---
title: "GEEutils: reproducing the pseudoreplication paper by Zimmerman et al."
output: rmarkdown::html_vignette
author: "Jeroen Gilis"
date: "09/01/2020"
vignette: >
  %\VignetteIndexEntry{Zimmerman_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Here I reproduce the analysis strategies of the publication by [Zimmerman](https://www.biorxiv.org/content/10.1101/2020.01.15.906248v2) et al.; Pseudoreplication bias in single-cell studies; a practical solution. See also their associated [GitHub page](https://github.com/kdzimm/PseudoreplicationPaper).

More specifically, I have included three functions in GEEutils:

- `getNoCor`: Only retain a select number of genes from a scRNA-Seq expression matrix that are not correlate to one another

- `getSimulationParameters`: Obtain hyperparameters describing an input scRNA-Seq dataset

- `simulator`: Simulate synthetic scRNA-Seq data based on hyperparameters obtained from a reference scRNA-Seq dataset with `getSimulationParameters`

Note that this code is almost identical to the one presented by Zimmerman et al. However, these authors assessed their methods only different types of scRNA-Seq data, i.e., Fluidigm's C1 protocol and the smartseq2 protocol. I will here work with the 10X Kang dataset. This has strong repercussions on the validity of the proposed methods. Therefore, adaptations must certainly be made in the future.

```
devtools::install_local("/Users/jg/Desktop/PhD/GEE/GEEutils", force=TRUE, quiet = FALSE)
```

# Load libraries

```{r, message=FALSE, warning=FALSE}
suppressPackageStartupMessages({
    library(ExperimentHub)
    library(magrittr)
    library(dplyr)
    library(scater)
    library(gdata)
})
```

# Loading the Kang data from ExperimentHub

```{r, message=FALSE}
eh <- ExperimentHub()
(sce <- eh[["EH2259"]])
```

# Data preprocessing

```{r}
# remove multiplets
sce <- sce[, sce$multiplets == "singlet" & !is.na(sce$cell)]

# Arrange colData
colData(sce) %>%
   as.data.frame %>%
   transmute(
       group_id = stim,
       patient_id = ind,
       sample_id = paste0(stim, ind),
       cluster_id = cell) %>%
   mutate_all(as.factor) %>%
   set_rownames(colnames(sce)) %>%
   DataFrame -> colData(sce)

# remove undetected genes
sce[rowSums(counts(sce) > 0) > 0, ]

# calculate quality control (QC) metrics
sce <- addPerCellQC(sce)

# get cells w/ few/many detected genes
sce$is_outlier <- isOutlier(
   metric = sce$detected,
   nmads = 2, type = "both", log = TRUE)

# remove outlier cells
sce <- sce[, !sce$is_outlier]
dim(sce)

# remove lowly expressed genes & normalize
sce <- sce[rowSums(counts(sce) > 1) >= 10, ]
sizeFactors(sce) <- librarySizeFactors(sce)
sce <- logNormCounts(sce)
```

Take data from a single celltype in a single treatment arm (but for multiple individuals) and only 1 cell type. Filter features again.

```{r}
sce <- sce[,sce$group_id == "ctrl"]
sce <- sce[,sce$cluster_id=="CD14+ Monocytes"]
(sce <- sce[rowSums(counts(sce) > 1) >= 50, ])
```

# Pull uncorrelated genes

```{r}
sceNoCor <- GEEutils::getNoCor(object = sce,
                               nFeatures = 500,
                               method = "spearman",
                               cutoff = 0.25)
```

# Run a tSNE plot

```{r}
GEEutils::runDimRed(object=sceNoCor,
                    clustering = "patient_id",
                    type = "TSNE",
                    downsample = FALSE)
```

# Assess intra- and inter-individual correlation with boxplots

```{r}
gg <- GEEutils::correlationsBoxplot(object = sceNoCor,
                                    clustering = "patient_id")
gg
```

# Get the simulation hyperparameters

```{r}
SimulationParameters <- GEEutils::getSimulationParameters(object = sceNoCor,
                                                clustering = "patient_id",
                                                plot = TRUE,
                                                verbose = TRUE)

SimulationParameters

# SimulationParameters <- GEEutils::getSimulationParameters(object = sce,
#                                                 clustering = "patient_id",
#                                                 plot = TRUE,
#                                                 verbose = TRUE)
# SimulationParameters
```

# Prepare to simulate new data

```{r}
ncells.group1 <- c(136, 176, 644, 315 ) # exact same cell numbers per individual
ncells.group2 <- c(86, 314, 292, 251) # exact same cell numbers per individual
# ncells.group1 <- rep(200, 5) # 5 patients with 200 cells in group 1
# ncells.group2 <- rep(200, 5) # 5 patients with 200 cells in group 2
ncells <- c(ncells.group1, ncells.group2)
cluster  <- paste0("patient_",1:8) # 10 patients in total
group <- rep(c("group1", "group2"), each=4)

design <- as.data.frame(cbind(cluster, group, ncells))
design$cluster <- as.factor(design$cluster)
design$group <- as.factor(design$group)
design$ncells <- as.numeric(design$ncells)

nGenes <- nrow(sce)
distribution <- "poisson" # alternative is NB
fractionDGE <- 0
foldchange <- 1
# fractionDGE <- .1
# foldchange <- 1.5

sum(ncells)
```

# Simulate new data

```{r}
# simulate from poisson
simulatedData_pois <- GEEutils::simulator(design = design,
                                SimulationParameters = SimulationParameters, 
                                nGenes = nGenes,
                                distribution = distribution,
                                fractionDGE = fractionDGE,
                                foldchange = foldchange)

simulatedData_pois[1:5,1:5]

# set.seed(55)
# simulatedData_NB <- GEEutils::simulator(design = design,
#                               SimulationParameters = SimulationParameters, 
#                               nGenes = nGenes,
#                               distribution = distribution,
#                               fractionDGE = fractionDGE,
#                               foldchange = foldchange)
# 
# simulatedData_NB[1:5,1:5]
# 
# sum(is.na(simulatedData_NB))
# simulatedData_NB[is.na(simulatedData_NB)] <- 0 # not nice, resolve later
```

# Make a countSimQC report of new data

```{r, eval = FALSE}
## original dataset
ids <- as.data.frame(cbind(colnames(sce), sce$patient_id))
colnames(ids) <- c("CellID", "IndividualID")
dds1 <- DESeqDataSetFromMatrix(countData = assay(sce),
                              colData = ids,
                              design= ~ IndividualID)

## dataset without gene correlations
ids <- as.data.frame(cbind(colnames(sceNoCor), sce$patient_id))
colnames(ids) <- c("CellID", "IndividualID")
dds2 <- DESeqDataSetFromMatrix(countData = assay(sceNoCor),
                              colData = ids,
                              design= ~ IndividualID)

## simulated dataset
cluster <- rep(design$cluster,times=design$ncells)
group <- rep(design$group,times=design$ncells)

designCell <- as.data.frame(cbind(cluster, group))
designCell$cluster <- as.factor(designCell$cluster)
designCell$group <- as.factor(designCell$group)

dds3 <- DESeqDataSetFromMatrix(countData = simulatedData_NB,
                              colData = designCell,
                              design= ~ cluster)

dds4 <- DESeqDataSetFromMatrix(countData = simulatedData_pois,
                              colData = designCell,
                              design= ~ cluster)


dds_list <- list(dds1, dds2, dds3, dds4)

names(dds_list) <- c("Kang_CD14_filtered", "Kang_CD14_filtered_noCor", "SimulatedData_NB", "SimulatedData_pois")

countsimQCReport(ddsList = dds_list, 
                 outputFile = "countsim_report_kang_simulation.html",
                 outputDir = getwd(), 
                 outputFormat = "html_document",
                 calculateStatistics = FALSE,
                 forceOverwrite = TRUE)
```
