<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GEEutils: reproducing the pseudoreplication paper by Zimmerman et al. • GEEutils</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="GEEutils: reproducing the pseudoreplication paper by Zimmerman et al.">
<meta property="og:description" content="GEEutils">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GEEutils</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Zimmerman_vignette.html">GEEutils: reproducing the pseudoreplication paper by Zimmerman et al.</a>
    </li>
    <li>
      <a href="../articles/bakery_vignette.html">GEEutils: a bakery for extra sandwiches</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jgilis/GEEutils/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Zimmerman_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>GEEutils: reproducing the pseudoreplication paper by Zimmerman et al.</h1>
                        <h4 class="author">Jeroen Gilis</h4>
            
            <h4 class="date">09/01/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jgilis/GEEutils/blob/master/vignettes/Zimmerman_vignette.Rmd"><code>vignettes/Zimmerman_vignette.Rmd</code></a></small>
      <div class="hidden name"><code>Zimmerman_vignette.Rmd</code></div>

    </div>

    
    
<p>Here I reproduce the analysis strategies of the publication by <a href="https://www.biorxiv.org/content/10.1101/2020.01.15.906248v2">Zimmerman</a> et al.; Pseudoreplication bias in single-cell studies; a practical solution. See also their associated <a href="https://github.com/kdzimm/PseudoreplicationPaper">GitHub page</a>.</p>
<p>More specifically, I have included three functions in GEEutils:</p>
<ul>
<li><p><code>getNoCor</code>: Only retain a select number of genes from a scRNA-Seq expression matrix that are not correlate to one another</p></li>
<li><p><code>getSimulationParameters</code>: Obtain hyperparameters describing an input scRNA-Seq dataset</p></li>
<li><p><code>simulator</code>: Simulate synthetic scRNA-Seq data based on hyperparameters obtained from a reference scRNA-Seq dataset with <code>getSimulationParameters</code></p></li>
</ul>
<p>Note that this code is almost identical to the one presented by Zimmerman et al. However, these authors assessed their methods only different types of scRNA-Seq data, i.e., Fluidigm’s C1 protocol and the smartseq2 protocol. I will here work with the 10X Kang dataset. This has strong repercussions on the validity of the proposed methods. Therefore, adaptations must certainly be made in the future.</p>
<pre><code>devtools::install_local("/Users/jg/Desktop/PhD/GEE/GEEutils", force=TRUE, quiet = FALSE)</code></pre>
<div id="load-libraries" class="section level1">
<h1 class="hasAnchor">
<a href="#load-libraries" class="anchor"></a>Load libraries</h1>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">gdata</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
</div>
<div id="loading-the-kang-data-from-experimenthub" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-the-kang-data-from-experimenthub" class="anchor"></a>Loading the Kang data from ExperimentHub</h1>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">(</span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH2259"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 35635 29065 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(35635): MIR1302-10 FAM138A ... MT-ND6 MT-CYB</span>
<span class="co">#&gt; rowData names(2): ENSEMBL SYMBOL</span>
<span class="co">#&gt; colnames(29065): AAACATACAATGCC-1 AAACATACATTTCC-1 ... TTTGCATGGTTTGG-1</span>
<span class="co">#&gt;   TTTGCATGTCTTAC-1</span>
<span class="co">#&gt; colData names(5): ind stim cluster cell multiplets</span>
<span class="co">#&gt; reducedDimNames(1): TSNE</span>
<span class="co">#&gt; altExpNames(0):</span></code></pre></div>
</div>
<div id="data-preprocessing" class="section level1">
<h1 class="hasAnchor">
<a href="#data-preprocessing" class="anchor"></a>Data preprocessing</h1>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># remove multiplets</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">multiplets</span> <span class="op">==</span> <span class="st">"singlet"</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span><span class="op">]</span>

<span class="co"># Arrange colData</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">%&gt;%</span>
   <span class="va">as.data.frame</span> <span class="op">%&gt;%</span>
   <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">transmute</a></span><span class="op">(</span>
       group_id <span class="op">=</span> <span class="va">stim</span>,
       patient_id <span class="op">=</span> <span class="va">ind</span>,
       sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">stim</span>, <span class="va">ind</span><span class="op">)</span>,
       cluster_id <span class="op">=</span> <span class="va">cell</span><span class="op">)</span> <span class="op">%&gt;%</span>
   <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html">mutate_all</a></span><span class="op">(</span><span class="va">as.factor</span><span class="op">)</span> <span class="op">%&gt;%</span>
   <span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html">set_rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
   <span class="va">DataFrame</span> <span class="op">-&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>

<span class="co"># remove undetected genes</span>
<span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 18464 24673 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(18464): RP11-34P13.8 AL627309.1 ... S100B PRMT2</span>
<span class="co">#&gt; rowData names(2): ENSEMBL SYMBOL</span>
<span class="co">#&gt; colnames(24673): AAACATACATTTCC-1 AAACATACCAGAAA-1 ... TTTGCATGGGACGA-1</span>
<span class="co">#&gt;   TTTGCATGTCTTAC-1</span>
<span class="co">#&gt; colData names(4): group_id patient_id sample_id cluster_id</span>
<span class="co">#&gt; reducedDimNames(1): TSNE</span>
<span class="co">#&gt; altExpNames(0):</span>

<span class="co"># calculate quality control (QC) metrics</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQC.html">addPerCellQC</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>

<span class="co"># get cells w/ few/many detected genes</span>
<span class="va">sce</span><span class="op">$</span><span class="va">is_outlier</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html">isOutlier</a></span><span class="op">(</span>
   metric <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">detected</span>,
   nmads <span class="op">=</span> <span class="fl">2</span>, type <span class="op">=</span> <span class="st">"both"</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># remove outlier cells</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">sce</span><span class="op">$</span><span class="va">is_outlier</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="co">#&gt; [1] 35635 22221</span>

<span class="co"># remove lowly expressed genes &amp; normalize</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span>, <span class="op">]</span>
<span class="fu">sizeFactors</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/librarySizeFactors.html">librarySizeFactors</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<p>Take data from a single celltype in a single treatment arm (but for multiple individuals) and only 1 cell type. Filter features again.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="va">sce</span><span class="op">$</span><span class="va">group_id</span> <span class="op">==</span> <span class="st">"ctrl"</span><span class="op">]</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="va">sce</span><span class="op">$</span><span class="va">cluster_id</span><span class="op">==</span><span class="st">"CD14+ Monocytes"</span><span class="op">]</span>
<span class="op">(</span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">50</span>, <span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 991 2214 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(2): counts logcounts</span>
<span class="co">#&gt; rownames(991): ISG15 AURKAIP1 ... CSTB SUMO3</span>
<span class="co">#&gt; rowData names(2): ENSEMBL SYMBOL</span>
<span class="co">#&gt; colnames(2214): AAACATACATTTCC-1 AAACATACCAGAAA-1 ... TTTCGAACTTGCTT-1</span>
<span class="co">#&gt;   TTTGACTGTTACCT-1</span>
<span class="co">#&gt; colData names(9): group_id patient_id ... is_outlier sizeFactor</span>
<span class="co">#&gt; reducedDimNames(1): TSNE</span>
<span class="co">#&gt; altExpNames(0):</span></code></pre></div>
</div>
<div id="pull-uncorrelated-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#pull-uncorrelated-genes" class="anchor"></a>Pull uncorrelated genes</h1>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sceNoCor</span> <span class="op">&lt;-</span> <span class="fu">GEEutils</span><span class="fu">::</span><span class="fu"><a href="../reference/getNoCor.html">getNoCor</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>,
                               nFeatures <span class="op">=</span> <span class="fl">500</span>,
                               method <span class="op">=</span> <span class="st">"spearman"</span>,
                               cutoff <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-a-tsne-plot" class="section level1">
<h1 class="hasAnchor">
<a href="#run-a-tsne-plot" class="anchor"></a>Run a tSNE plot</h1>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">GEEutils</span><span class="fu">::</span><span class="fu"><a href="../reference/runDimRed.html">runDimRed</a></span><span class="op">(</span>object<span class="op">=</span><span class="va">sceNoCor</span>,
                    clustering <span class="op">=</span> <span class="st">"patient_id"</span>,
                    type <span class="op">=</span> <span class="st">"TSNE"</span>,
                    downsample <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="Zimmerman_vignette_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div id="assess-intra--and-inter-individual-correlation-with-boxplots" class="section level1">
<h1 class="hasAnchor">
<a href="#assess-intra--and-inter-individual-correlation-with-boxplots" class="anchor"></a>Assess intra- and inter-individual correlation with boxplots</h1>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gg</span> <span class="op">&lt;-</span> <span class="fu">GEEutils</span><span class="fu">::</span><span class="fu"><a href="../reference/correlationsBoxplot.html">correlationsBoxplot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sceNoCor</span>,
                                    clustering <span class="op">=</span> <span class="st">"patient_id"</span><span class="op">)</span>
<span class="co">#&gt; Function still under construction</span>
<span class="va">gg</span></code></pre></div>
<p><img src="Zimmerman_vignette_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div id="get-the-simulation-hyperparameters" class="section level1">
<h1 class="hasAnchor">
<a href="#get-the-simulation-hyperparameters" class="anchor"></a>Get the simulation hyperparameters</h1>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SimulationParameters</span> <span class="op">&lt;-</span> <span class="fu">GEEutils</span><span class="fu">::</span><span class="fu"><a href="../reference/getSimulationParameters.html">getSimulationParameters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sceNoCor</span>,
                                                clustering <span class="op">=</span> <span class="st">"patient_id"</span>,
                                                plot <span class="op">=</span> <span class="cn">TRUE</span>,
                                                verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Summary statistics</span>
<span class="co">#&gt; Hyperparameters dropout</span>
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span>
<span class="co">#&gt; Hyperparameters grand mean</span></code></pre></div>
<p><img src="Zimmerman_vignette_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<pre><code>#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="Zimmerman_vignette_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
<pre><code>#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
#&gt; Warning: Removed 25 rows containing non-finite values (stat_bin).
#&gt; Warning: Removed 2 rows containing missing values (geom_bar).
#&gt; Hyperparameters realtionship 1</code></pre>
<p><img src="Zimmerman_vignette_files/figure-html/unnamed-chunk-9-3.png" width="700"><img src="Zimmerman_vignette_files/figure-html/unnamed-chunk-9-4.png" width="700"></p>
<pre><code>#&gt; Warning: Removed 36 rows containing non-finite values (stat_smooth).
#&gt; Warning: Removed 36 rows containing missing values (geom_point).
#&gt; Hyperparameters realtionship 2</code></pre>
<p><img src="Zimmerman_vignette_files/figure-html/unnamed-chunk-9-5.png" width="700"><img src="Zimmerman_vignette_files/figure-html/unnamed-chunk-9-6.png" width="700"></p>
<pre><code>#&gt; Warning: Removed 18 rows containing non-finite values (stat_smooth).
#&gt; Warning: Removed 18 rows containing missing values (geom_point).
#&gt; Combine all hyperparameters</code></pre>
<p><img src="Zimmerman_vignette_files/figure-html/unnamed-chunk-9-7.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">SimulationParameters</span>
<span class="co">#&gt; $drop.shape</span>
<span class="co">#&gt; [1] 4.101524</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $drop.rate</span>
<span class="co">#&gt; [1] 12.17911</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $drop.beta0</span>
<span class="co">#&gt; [1] 0.01257415</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $drop.beta1</span>
<span class="co">#&gt; [1] 0.1895456</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $drop.beta2</span>
<span class="co">#&gt; [1] -0.1819284</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $grandmean.shape</span>
<span class="co">#&gt; [1] 3.470282</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $grandmean.rate</span>
<span class="co">#&gt; [1] 2.056208</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $grandmean_interSD.beta1</span>
<span class="co">#&gt; [1] 0.1215916</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $intraSD_dispersion.beta0</span>
<span class="co">#&gt; [1] 2.762912</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $intraSD_dispersion.beta1</span>
<span class="co">#&gt; [1] 0.5880615</span>

<span class="co"># SimulationParameters &lt;- GEEutils::getSimulationParameters(object = sce,</span>
<span class="co">#                                                 clustering = "patient_id",</span>
<span class="co">#                                                 plot = TRUE,</span>
<span class="co">#                                                 verbose = TRUE)</span>
<span class="co"># SimulationParameters</span></code></pre></div>
</div>
<div id="prepare-to-simulate-new-data" class="section level1">
<h1 class="hasAnchor">
<a href="#prepare-to-simulate-new-data" class="anchor"></a>Prepare to simulate new data</h1>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ncells.group1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">136</span>, <span class="fl">176</span>, <span class="fl">644</span>, <span class="fl">315</span> <span class="op">)</span> <span class="co"># exact same cell numbers per individual</span>
<span class="va">ncells.group2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">86</span>, <span class="fl">314</span>, <span class="fl">292</span>, <span class="fl">251</span><span class="op">)</span> <span class="co"># exact same cell numbers per individual</span>
<span class="co"># ncells.group1 &lt;- rep(200, 5) # 5 patients with 200 cells in group 1</span>
<span class="co"># ncells.group2 &lt;- rep(200, 5) # 5 patients with 200 cells in group 2</span>
<span class="va">ncells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">ncells.group1</span>, <span class="va">ncells.group2</span><span class="op">)</span>
<span class="va">cluster</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"patient_"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span> <span class="co"># 10 patients in total</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"group1"</span>, <span class="st">"group2"</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">4</span><span class="op">)</span>

<span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">cluster</span>, <span class="va">group</span>, <span class="va">ncells</span><span class="op">)</span><span class="op">)</span>
<span class="va">design</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">design</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>
<span class="va">design</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">design</span><span class="op">$</span><span class="va">group</span><span class="op">)</span>
<span class="va">design</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">design</span><span class="op">$</span><span class="va">ncells</span><span class="op">)</span>

<span class="va">nGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">distribution</span> <span class="op">&lt;-</span> <span class="st">"poisson"</span> <span class="co"># alternative is NB</span>
<span class="va">fractionDGE</span> <span class="op">&lt;-</span> <span class="fl">0</span>
<span class="va">foldchange</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="co"># fractionDGE &lt;- .1</span>
<span class="co"># foldchange &lt;- 1.5</span>

<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">ncells</span><span class="op">)</span>
<span class="co">#&gt; [1] 2214</span></code></pre></div>
</div>
<div id="simulate-new-data" class="section level1">
<h1 class="hasAnchor">
<a href="#simulate-new-data" class="anchor"></a>Simulate new data</h1>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># simulate from poisson</span>
<span class="va">simulatedData_pois</span> <span class="op">&lt;-</span> <span class="fu">GEEutils</span><span class="fu">::</span><span class="fu"><a href="../reference/simulator.html">simulator</a></span><span class="op">(</span>design <span class="op">=</span> <span class="va">design</span>,
                                SimulationParameters <span class="op">=</span> <span class="va">SimulationParameters</span>, 
                                nGenes <span class="op">=</span> <span class="va">nGenes</span>,
                                distribution <span class="op">=</span> <span class="va">distribution</span>,
                                fractionDGE <span class="op">=</span> <span class="va">fractionDGE</span>,
                                foldchange <span class="op">=</span> <span class="va">foldchange</span><span class="op">)</span>

<span class="va">simulatedData_pois</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;        Cell_1 Cell_2 Cell_3 Cell_4 Cell_5</span>
<span class="co">#&gt; Gene_1      0      7      9      0      0</span>
<span class="co">#&gt; Gene_2      1      3      0      2      2</span>
<span class="co">#&gt; Gene_3      4      0      0      0      0</span>
<span class="co">#&gt; Gene_4      2      3      0      0      0</span>
<span class="co">#&gt; Gene_5      0      0      0      0      0</span>

<span class="co"># set.seed(55)</span>
<span class="co"># simulatedData_NB &lt;- GEEutils::simulator(design = design,</span>
<span class="co">#                               SimulationParameters = SimulationParameters, </span>
<span class="co">#                               nGenes = nGenes,</span>
<span class="co">#                               distribution = distribution,</span>
<span class="co">#                               fractionDGE = fractionDGE,</span>
<span class="co">#                               foldchange = foldchange)</span>
<span class="co"># </span>
<span class="co"># simulatedData_NB[1:5,1:5]</span>
<span class="co"># </span>
<span class="co"># sum(is.na(simulatedData_NB))</span>
<span class="co"># simulatedData_NB[is.na(simulatedData_NB)] &lt;- 0 # not nice, resolve later</span></code></pre></div>
</div>
<div id="make-a-countsimqc-report-of-new-data" class="section level1">
<h1 class="hasAnchor">
<a href="#make-a-countsimqc-report-of-new-data" class="anchor"></a>Make a countSimQC report of new data</h1>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## original dataset</span>
<span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">sce</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CellID"</span>, <span class="st">"IndividualID"</span><span class="op">)</span>
<span class="va">dds1</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,
                              colData <span class="op">=</span> <span class="va">ids</span>,
                              design<span class="op">=</span> <span class="op">~</span> <span class="va">IndividualID</span><span class="op">)</span>

<span class="co">## dataset without gene correlations</span>
<span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sceNoCor</span><span class="op">)</span>, <span class="va">sce</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CellID"</span>, <span class="st">"IndividualID"</span><span class="op">)</span>
<span class="va">dds2</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sceNoCor</span><span class="op">)</span>,
                              colData <span class="op">=</span> <span class="va">ids</span>,
                              design<span class="op">=</span> <span class="op">~</span> <span class="va">IndividualID</span><span class="op">)</span>

<span class="co">## simulated dataset</span>
<span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="va">design</span><span class="op">$</span><span class="va">cluster</span>,times<span class="op">=</span><span class="va">design</span><span class="op">$</span><span class="va">ncells</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="va">design</span><span class="op">$</span><span class="va">group</span>,times<span class="op">=</span><span class="va">design</span><span class="op">$</span><span class="va">ncells</span><span class="op">)</span>

<span class="va">designCell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">cluster</span>, <span class="va">group</span><span class="op">)</span><span class="op">)</span>
<span class="va">designCell</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">designCell</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>
<span class="va">designCell</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">designCell</span><span class="op">$</span><span class="va">group</span><span class="op">)</span>

<span class="va">dds3</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">simulatedData_NB</span>,
                              colData <span class="op">=</span> <span class="va">designCell</span>,
                              design<span class="op">=</span> <span class="op">~</span> <span class="va">cluster</span><span class="op">)</span>

<span class="va">dds4</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span><span class="op">(</span>countData <span class="op">=</span> <span class="va">simulatedData_pois</span>,
                              colData <span class="op">=</span> <span class="va">designCell</span>,
                              design<span class="op">=</span> <span class="op">~</span> <span class="va">cluster</span><span class="op">)</span>


<span class="va">dds_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">dds1</span>, <span class="va">dds2</span>, <span class="va">dds3</span>, <span class="va">dds4</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">dds_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Kang_CD14_filtered"</span>, <span class="st">"Kang_CD14_filtered_noCor"</span>, <span class="st">"SimulatedData_NB"</span>, <span class="st">"SimulatedData_pois"</span><span class="op">)</span>

<span class="fu">countsimQCReport</span><span class="op">(</span>ddsList <span class="op">=</span> <span class="va">dds_list</span>, 
                 outputFile <span class="op">=</span> <span class="st">"countsim_report_kang_simulation.html"</span>,
                 outputDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span>, 
                 outputFormat <span class="op">=</span> <span class="st">"html_document"</span>,
                 calculateStatistics <span class="op">=</span> <span class="cn">FALSE</span>,
                 forceOverwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jeroen Gilis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
